src/vuln_analysis/utils/source_rpm_downloader.py

```diff
@@ -0,0 +1,731 @@
+   #     # BFS to depth
+   #     needed = set()
+   #     queue = deque([(project_name, 0)])
+   #     while queue:
+   #         node, d = queue.popleft()
+   #         if node in needed or d > depth:
+   #             continue
+   #         needed.add(node)
+   #         if node in dependency_tree:
+   #             for child in dependency_tree[node]:
+   #                 queue.append((child, d + 1))
+   #     # List all RPM dirs
+   #     rpm_libs = Path(rpm_libs_path)
+   #     all_rpm_dirs = [p.name for p in rpm_libs.iterdir() if p.is_dir()]
+        # Prune: keep only those in needed
+   #     to_delete = [d for d in all_rpm_dirs if d not in needed]        
+   #     logger.info(f"RPMs that can be deleted (not needed for analysis): {len(to_delete)} from total {len(all_rpm_dirs)}")
+        
+   #     # Delete unused RPM directories
+   #     for d in to_delete:
+   #         dir_path = rpm_libs / d
+   #         try:
+   #             shutil.rmtree(dir_path)                
+   #         except Exception as e:
+   #             logger.error(f"Failed to delete {dir_path}: {e}")
+        # Remove deleted elements from the dependency tree
+   #     updated_tree = {k: [v for v in vs if v not in to_delete] for k, vs in dependency_tree.items() if k not in to_delete}
+   #     return updated_tree
+
+   # def write_output(self, output_path, data):
+   #     with open(output_path, 'w') as f:
+   #         json.dump(data, f, indent=2)
+
+
+class SourceRPMDownloader:
+    def __init__(self, prj_name, sbom_info, repo_urls: List[RepoUrl], work_dir="ext_libs", rpm_cache_dir=None, max_workers=1):
+        self.sbom_packages = sbom_info
+        self.repo_urls = repo_urls  # Now a list of RepoUrl objects
+        self.work_dir = Path(work_dir)
+        
+        # Set up RPM cache directory
+        if rpm_cache_dir is None:
+            # No cache directory specified - disable caching
+            self.rpm_cache_dir = None
+            logger.info("RPM caching disabled (no cache directory specified)")
+        else:
+            self.rpm_cache_dir = Path(rpm_cache_dir)
+            # Ensure cache directory exists
+            self.rpm_cache_dir.mkdir(parents=True, exist_ok=True)
+            logger.debug(f"RPM cache directory: {self.rpm_cache_dir}")
+        
+        # Build cache index from existing files
+        self.cache_index = self._build_cache_index()
+        self.cache_hit =0
+        self.cache_miss =0
+        self.repo_metadata = {}
+        self.max_workers = max_workers
+        self.project_name = prj_name
+        # Thread-safe session with retries
+        self.session = requests.Session()
+        retries = Retry(
+            total=3, backoff_factor=0.3, status_forcelist=[500, 502, 503, 504], allowed_methods=["GET"]
+        )
+        adapter = HTTPAdapter(max_retries=retries)
+        self.session.mount("http://", adapter)
+        self.session.mount("https://", adapter)
+
+    def _parse_rpm_filename(self, filename):
```

Comment:
@RedTanny Please use for method parameters type hints/annotations for an increased readability and maintainability.